feat(erc-xxxx): add on-chain timestamps to ONCHAINID claims

SaschaKubisch opened the pull request:
- Add createdAt and updatedAt uint64 fields to Claim struct
- Extend getClaim() to return both timestamp values
- Set createdAt immutably on first claim creation
- Update updatedAt on every claim modification
- Use block.timestamp for authoritative on-chain timing

Timestamps enable claim freshness verification and audit trails directly onchain without relying on events or external indexers. The timestamps are portable across chains as they travel with the claim data itself.

SaschaKubisch committed:
feat(erc-xxxx): add on-chain timestamps to ONCHAINID claims

- Add createdAt and updatedAt uint64 fields to Claim struct
- Extend getClaim() to return both timestamp values
- Set createdAt immutably on first claim creation
- Update updatedAt on every claim modification
- Use block.timestamp for authoritative on-chain timing

Timestamps enable claim freshness verification and audit trails without
relying on events or external indexers. The timestamps are portable
across chains as they travel with the claim data itself.

Implementation preserves the original single-assignment pattern and
maintains backward compatibility except for the intentional getClaim
signature change.

Joachim-Lebrun commented:
interesting approach, i had in mind to use the data field to store these values as you can see in the specification document, the interest being that the data field is part of what the claim issuer signs, so it is proven, while the createdAt and updatedAt only give information about when the claim was added/updated onchain. if the identity owner has the permission to remove claims they could as well bypass the updatedAt part by removing then adding again the same claim and it would look like brand new, right?

Joachim-Lebrun commented:
instead of this we could probably standardize the claim data field to say that for every claim the first encoded parameter will always be the timestamp at which the claim issuer validated the claim and that parameter to the getClaim function, potentially with an expiry date too, that could be the second parameter of the data field, and we can standardize it in a way that if the expiry is 0 it means the claim is valid forever (no expiry) or something like that

SaschaKubisch commented:
I didnt think about that, but you are right, this should be technically possible. 

I am trying to understand in which cases this would be in the interest of a malicious actor.
Why would the identity owner not want to update but instead issue a new claim? And who would be the identity owner? Is it the instance that deploys the identity contract / the owner of the infra? Why would they be interested in tampering their own system? Just trying to get a better sense of how the system would be used IRL.

SaschaKubisch commented:
And adding the timestamp + expiry standardized into the data field makes sense to have it completely
  tamperproof.

  I am thinking of how to make it easy to handle. 

Should we add a helper library for claim issuers to encode timestamp + expiry + data?

  Something like:

```
  library ClaimDataEncoder {

      function encodeClaimWithExpiry(
          uint256 expiryTimestamp,
          bytes memory claimData
      ) internal view returns (bytes memory) {
          return abi.encode(
              block.timestamp,     // issuedAt
              expiryTimestamp,     // validUntil
              claimData
          );
      }
  }
```

  This would standardize the data field as: [issuedAt, validUntil, claimData] where validUntil = 0 means no
   expiry.

SaschaKubisch commented:
Alternatively, we could make it even simpler for issuers by having the helper calculate the expiry from a
   validity period:

```
  library ClaimDataEncoder {
      // Validity period in seconds (auto-calculates expiry)
      function encodeClaimWithValidity(
          bytes memory claimData,
          uint256 validitySeconds
      ) internal view returns (bytes memory) {
          return abi.encode(
              block.timestamp,                      // issuedAt
              block.timestamp + validitySeconds,    // validUntil (calculated)
              claimData
          );
      }

      // No expiry (permanent claims)
      function encodeClaim(
          bytes memory claimData
      ) internal view returns (bytes memory) {
          return abi.encode(
              block.timestamp,     // issuedAt
              0,                   // validUntil (0 = no expiry)
              claimData
          );
      }
  }

```
  Usage would be:
  // KYC valid for 180 days
  bytes memory data = encodeClaimWithValidity(kycData, 180 days);

  // Permanent nationality claim
  bytes memory data = encodeClaim(nationalityData);

  This makes use of Solidity's built-in time units (days, hours, etc.) and removes the need for issuers to
  calculate Unix timestamps.

Joachim-Lebrun commented:
> Should we add a helper library for claim issuers to encode timestamp + expiry + data?

that can be interesting indeed, it can be managed on the app side also otherwise

Joachim-Lebrun commented:
I think you can update your PR by adding this standardized approach to store the claim timestamp and expiry within the data field, this information will be useful for any claim i believe.

SaschaKubisch committed:
feat: standardize claim timestamps within data field

  Move timestamps into the data field to make them tamper-proof and part of
  the signed claim data, addressing review feedback.

  - Add ClaimDataHelper library for standardized timestamp encoding/decoding
  - Remove createdAt/updatedAt from Claim struct and getClaim interface
  - Encode timestamps as [issuedAt, validUntil, claimData] in data field
  - Support permanent claims with validUntil=0 and time-limited claims
  - Add usage examples for KYC claim issuance and verification
  - Update ClaimIssuer to validate expiry using the helper library

  This ensures timestamps cannot be manipulated after claim issuance since
  they become part of what the issuer cryptographically signs.

SaschaKubisch commented:
> I think you can update your PR by adding this standardized approach to store the claim timestamp and expiry within the data field, this information will be useful for any claim i believe.

Done @Joachim-Lebrun
